### DISCUSSION TITLE
Generic Smart Curtain new device

### DISCUSSION THREAD

**Discussion by wildchai** [Koenkk/zigbee2mqtt#18985] [Devices]
Hi,

I just got a generic smart curtain from taobao. It connects to Z2M but it was unsupported, so I followed the guide. Managed to get it supported but was unable to properly set the converters. Tried to reuse existing converters but still the same error.

**Debug 2023-09-16 15:40:21Received MQTT message on 'zigbee2mqtt/Curtain/set' with data '{"state":"OPEN"}'
Error 2023-09-16 15:40:21No converter available for 'state' ("OPEN")**

const fz = require('zigbee-herdsman-converters/converters/fromZigbee');
const tz = require('zigbee-herdsman-converters/converters/toZigbee');
const exposes = require('zigbee-herdsman-converters/lib/exposes');
const reporting = require('zigbee-herdsman-converters/lib/reporting');
const extend = require('zigbee-herdsman-converters/lib/extend');
const e = exposes.presets;
const ea = exposes.access;

const definition = {
    zigbeeModel: ['TS0601'], // The model ID from: Device with modelID 'lumi.sens' is not supported.
    model: '_TZE204_1fuxihti', // Vendor model number, look on the device for a model number
    vendor: 'Tuya', // Vendor of the device (only used for documentation and startup logging)
    description: 'Smart Curtain', // Description of the device, copy from vendor site. (only used for documentation and startup logging)
    fromZigbee: [fz.curtain_position_analog_output,fz.xiaomi_curtain_position], // We will add this later
    toZigbee: [], // Should be empty, unless device can be controlled (e.g. lights, switches).
    exposes: [e.cover_position()], // Defines what this device exposes, used for e.g. Home Assistant discovery and in the frontend
};

**Comment by wildchai** [comment]
any help with adding generic curtain motors?

**Comment by gbrest** [comment]
Hey everyone! 👋

Just wanted to share a bit of my experience with the **MOES Zigbee Curtain Motor** (`_TZE204_1fuxihti`, model `TS0601`) and how I finally got it working in **Home Assistant with ZHA**, after a lot of digging and frustration.

---

📆 **As of mid-2025, the issue is still ongoing**

I bought a couple of these curtain motors in 2025 and paired them to Home Assistant using ZHA. To get them into pairing mode, I:

* Pressed the reset button **three times** quickly
* Then held it down until the LED started flashing

The device paired and showed up in the ZHA device list, but — **no entities appeared.**

---

❗**Why? Because it’s not really Zigbee-native**

This is another case of a Tuya TS0601 device. It communicates using Tuya’s proprietary MCU protocol over the `0xEF00` cluster, which is **not natively understood by ZHA**.

That’s why the device shows up, but doesn’t expose any entities or usable functionality unless you provide a **custom quirk**.

I found an old discussion where user **Prozpect** had posted a working quirk, but the link is gone. I had to do a deep dive across GitHub, Reddit and the Wayback Machine to find pieces of it and put it all back together.

---

🧩 **So here’s the full working solution, in one place, for rookies like me**

This is everything I did to get the curtain working properly with ZHA, including open/close/stop and position slider.

---

 ✅ **Step 1: Create the quirks folder**

Inside your `/config` directory (where `configuration.yaml` lives), create a new folder called:

**quirks**

So the full path is:

/config/quirks/

---

 ✅ **Step 2: Create an empty `__init__.py`**

Inside the `quirks` folder, create a new file called:

__init__.py

It can be completely empty, but it **must exist**.

---

 ✅ **Step 3: Create the custom quirk file**

Still inside the `quirks` folder, create a file called:

moes_curtain.py

And paste the following code into it (this is the full working quirk):

```

"""Tuya MCU based cover and blinds."""

from typing import Dict, Optional, Union
from zigpy.profiles import zha
import zigpy.types as t
from zigpy.zcl import foundation
from zigpy.zcl.clusters.closures import WindowCovering
from zigpy.zcl.clusters.general import (
    Basic,
    GreenPowerProxy,
    Groups,
    Ota,
    Scenes,
    Time,
)

from zhaquirks.const import (
    DEVICE_TYPE,
    ENDPOINTS,
    INPUT_CLUSTERS,
    MODELS_INFO,
    OUTPUT_CLUSTERS,
    PROFILE_ID,
)
from zhaquirks.tuya import (
    NoManufacturerCluster,
    TUYA_MCU_COMMAND,
    TuyaLocalCluster,
    TuyaWindowCover,
    TuyaManufacturerWindowCover,
)
from zhaquirks.tuya.mcu import (
    DPToAttributeMapping,
    TuyaClusterData,
    TuyaMCUCluster,
)


class TuyaCC(t.enum8):
    """Tuya cover commands."""
    OPEN = 0x00
    STOP = 0x01
    CLOSE = 0x02


class ZclCC(t.enum8):
    """ZCL cover commands."""
    OPEN = 0x00
    CLOSE = 0x01
    STOP = 0x02


TUYA2ZB_COMMANDS = {
    ZclCC.OPEN: TuyaCC.OPEN,
    ZclCC.CLOSE: TuyaCC.CLOSE,
    ZclCC.STOP: TuyaCC.STOP,
}


class TuyaWindowCovering(NoManufacturerCluster, WindowCovering, TuyaLocalCluster):
    """Tuya MCU WindowCovering cluster."""

    attributes = WindowCovering.attributes.copy()
    attributes.update(
        {
            0xF000: ("curtain_switch", t.enum8, True),
            0xF001: ("accurate_calibration", t.enum8, True),
            0xF002: ("motor_steering", t.enum8, True),
            0xF003: ("travel", t.uint16_t, True),
        }
    )

    async def command(
        self,
        command_id: Union[foundation.GeneralCommand, int, t.uint8_t],
        *args,
        manufacturer: Optional[Union[int, t.uint16_t]] = None,
        expect_reply: bool = True,
        tsn: Optional[Union[int, t.uint8_t]] = None,
    ):
        """Override the default Cluster command."""

        self.debug(
            "Sending Tuya Cluster Command. Cluster Command is %x, Arguments are %s",
            command_id,
            args,
        )

        if command_id in (0x0002, 0x0000, 0x0001):
            cluster_data = TuyaClusterData(
                endpoint_id=self.endpoint.endpoint_id,
                cluster_name=self.ep_attribute,
                cluster_attr="curtain_switch",
                attr_value=TUYA2ZB_COMMANDS[command_id],
                expect_reply=expect_reply,
                manufacturer=-1,
            )
            self.endpoint.device.command_bus.listener_event(
                TUYA_MCU_COMMAND,
                cluster_data,
            )
            return foundation.GENERAL_COMMANDS[
                foundation.GeneralCommand.Default_Response
            ].schema(command_id=command_id, status=foundation.Status.SUCCESS)

        elif command_id == 0x0005:
            lift_value = args[0]

            cluster_data = TuyaClusterData(
                endpoint_id=self.endpoint.endpoint_id,
                cluster_name=self.ep_attribute,
                cluster_attr="current_position_lift_percentage",
                attr_value=lift_value,
                expect_reply=expect_reply,
                manufacturer=-1,
            )
            self.endpoint.device.command_bus.listener_event(
                TUYA_MCU_COMMAND,
                cluster_data,
            )
            return foundation.GENERAL_COMMANDS[
                foundation.GeneralCommand.Default_Response
            ].schema(command_id=command_id, status=foundation.Status.SUCCESS)

        self.warning("Unsupported command_id: %s", command_id)
        return foundation.GENERAL_COMMANDS[
            foundation.GeneralCommand.Default_Response
        ].schema(command_id=command_id, status=foundation.Status.UNSUP_CLUSTER_COMMAND)


class TuyaWindowCoverManufCluster(TuyaMCUCluster):
    """Tuya with WindowCover data points."""

    attributes = TuyaMCUCluster.attributes.copy()
    attributes.update(
        {
            0x5000: ("backlight_mode", t.enum8, True),
            0x8001: ("indicator_status", t.enum8, True),
        }
    )

    dp_to_attribute: Dict[int, DPToAttributeMapping] = {
        1: DPToAttributeMapping(
            TuyaWindowCovering.ep_attribute,
            "curtain_switch",
        ),
        2: DPToAttributeMapping(
            TuyaWindowCovering.ep_attribute,
            "current_position_lift_percentage",
        ),
        3: DPToAttributeMapping(
            TuyaWindowCovering.ep_attribute,
            "current_position_lift_percentage",
        ),
    }

    data_point_handlers = {
        1: "_dp_2_attr_update",
        2: "_dp_2_attr_update",
        3: "_dp_2_attr_update",
    }


class TuyaCover0601_GP(TuyaWindowCover):
    """Tuya blind controller device."""

    signature = {
        MODELS_INFO: [
            ("_TZE204_1fuxihti", "TS0601"),
        ],
        ENDPOINTS: {
            1: {
                PROFILE_ID: zha.PROFILE_ID,
                DEVICE_TYPE: zha.DeviceType.SMART_PLUG,
                INPUT_CLUSTERS: [
                    Basic.cluster_id,
                    Groups.cluster_id,
                    Scenes.cluster_id,
                    TuyaWindowCoverManufCluster.cluster_id,
                ],
                OUTPUT_CLUSTERS: [Time.cluster_id, Ota.cluster_id],
            },
            242: {
                PROFILE_ID: 41440,
                DEVICE_TYPE: 97,
                INPUT_CLUSTERS: [],
                OUTPUT_CLUSTERS: [GreenPowerProxy.cluster_id],
            },
        },
    }

    replacement = {
        ENDPOINTS: {
            1: {
                DEVICE_TYPE: zha.DeviceType.WINDOW_COVERING_DEVICE,
                INPUT_CLUSTERS: [
                    Basic.cluster_id,
                    Groups.cluster_id,
                    Scenes.cluster_id,
                    TuyaWindowCoverManufCluster,
                    TuyaWindowCovering,
                ],
                OUTPUT_CLUSTERS: [Time.cluster_id, Ota.cluster_id],
            },
            242: {
                PROFILE_ID: 41440,
                DEVICE_TYPE: 97,
                INPUT_CLUSTERS: [],
                OUTPUT_CLUSTERS: [GreenPowerProxy.cluster_id],
            },
        }
    }
```

---

✅ **Step 4: Tell ZHA to load custom quirks**

In your `configuration.yaml`, add or update the `zha:` section like this:

```
zha:
  custom_quirks_path: quirks
```

> **Do not** include `/config/quirks` — just `quirks`

---

 ✅ **Step 5: Restart and pair again**

1. Restart Home Assistant completely (not just ZHA).
2. Remove any previous pairing of the curtain motor.
3. Put the curtain back into pairing mode (3 presses, then long press).
4. Pair it again from **ZHA > Add Device**.

You should now see a new **`cover.xxxxxx` entity**, and you’ll be able to open/close/stop and use the position slider.

---

Let me know if this works for you or if you get stuck — happy to help!

**Reply by Steven-Low** [reply]
BROOOO, You are LEGEND. THANK YOU SO MUCH <3 
IT WORK 100% PERFECT

<img width="640" height="542" alt="image" src="https://github.com/user-attachments/assets/d45fd8d3-1871-4579-b3cd-3b01189d2213" />
